File: zhiguang_be/src/main/java/com/tongji/relation/service/RelationService.java
package com.tongji.relation.service;

import java.util.List;
import java.util.Map;
import com.tongji.profile.api.dto.ProfileResponse;

/**
 * 关系服务接口。
 * 能力：关注/取消关注、关系查询、关注/粉丝列表（偏移与游标），以及用户资料视图的聚合组装。
 */
public interface RelationService {
    /**
     * 关注操作。
     * @param fromUserId 发起关注的用户ID
     * @param toUserId 被关注的用户ID
     * @return 是否关注成功
     */
    boolean follow(long fromUserId, long toUserId);
    /**
     * 取消关注。
     * @param fromUserId 发起取消的用户ID
     * @param toUserId 被取消的用户ID
     * @return 是否取消成功
     */
    boolean unfollow(long fromUserId, long toUserId);
    /**
     * 判断是否已关注。
     * @param fromUserId 关注发起者
     * @param toUserId 被关注者
     * @return 是否已关注
     */
    boolean isFollowing(long fromUserId, long toUserId);
    /**
     * 获取关注列表（偏移分页）。
     * @param userId 用户ID
     * @param limit 返回数量上限
     * @param offset 偏移量
     * @return 关注用户ID列表
     */
    List<Long> following(long userId, int limit, int offset);
    /**
     * 获取粉丝列表（偏移分页）。
     * @param userId 用户ID
     * @param limit 返回数量上限
     * @param offset 偏移量
     * @return 粉丝用户ID列表
     */
    List<Long> followers(long userId, int limit, int offset);
    /**
     * 查询双方关系三态。
     * @param userId 当前用户ID
     * @param otherUserId 对方用户ID
     * @return following/followedBy/mutual 三态
     */
    Map<String, Boolean> relationStatus(long userId, long otherUserId);
    /**
     * 游标分页关注列表。
     * @param userId 用户ID
     * @param limit 返回数量上限
     * @param cursor 上一页末条的分数（毫秒时间戳），为空代表第一页
     * @return 关注用户ID列表
     */
    List<Long> followingCursor(long userId, int limit, Long cursor);
    /**
     * 游标分页粉丝列表。
     * @param userId 用户ID
     * @param limit 返回数量上限
     * @param cursor 上一页末条的分数（毫秒时间戳），为空代表第一页
     * @return 粉丝用户ID列表
     */
    List<Long> followersCursor(long userId, int limit, Long cursor);

    /**
     * 关注列表（资料视图）：在 ID 列表基础上聚合用户资料，支持偏移或游标。
     * @param userId 用户ID
     * @param limit 返回数量上限
     * @param offset 偏移量（游标为空时生效）
     * @param cursor 游标（上一页末条分数）
     * @return 关注用户的资料视图列表
     */
    List<ProfileResponse> followingProfiles(long userId, int limit, int offset, Long cursor);

    /**
     * 粉丝列表（资料视图）：在 ID 列表基础上聚合用户资料，支持偏移或游标。
     * @param userId 用户ID
     * @param limit 返回数量上限
     * @param offset 偏移量（游标为空时生效）
     * @param cursor 游标（上一页末条分数）
     * @return 粉丝用户的资料视图列表
     */
    List<ProfileResponse> followersProfiles(long userId, int limit, int offset, Long cursor);
}

File: yixiang-web/src/services/relationService.ts
import { apiFetch } from "@/lib/apiClient";
import type { RelationStatusResponse, RelationCountersResponse } from "@/types/relation";
import type { ProfileResponse } from "@/types/profile";

const RELATION_PREFIX = "/api/v1/relation";

export const relationService = {
  follow: (toUserId: number) =>
    apiFetch<boolean>(`${RELATION_PREFIX}/follow?toUserId=${toUserId}`, {
      method: "POST",
    }),

  unfollow: (toUserId: number) =>
    apiFetch<boolean>(`${RELATION_PREFIX}/unfollow?toUserId=${toUserId}`, {
      method: "POST",
    }),

  status: (toUserId: number) =>
    apiFetch<RelationStatusResponse>(`${RELATION_PREFIX}/status?toUserId=${toUserId}`),

  following: (userId: number, limit = 20, offset = 0, cursor?: number, accessToken?: string) => {
    const params = new URLSearchParams({ userId: String(userId), limit: String(limit), offset: String(offset) });
    if (typeof cursor === "number") params.set("cursor", String(cursor));
    return apiFetch<ProfileResponse[]>(`${RELATION_PREFIX}/following?${params.toString()}`, {
      skipAuth: !accessToken
    });
  },

  followers: (userId: number, limit = 20, offset = 0, cursor?: number, accessToken?: string) => {
    const params = new URLSearchParams({ userId: String(userId), limit: String(limit), offset: String(offset) });
    if (typeof cursor === "number") params.set("cursor", String(cursor));
    return apiFetch<ProfileResponse[]>(`${RELATION_PREFIX}/followers?${params.toString()}`, {
      skipAuth: !accessToken
    });
  },

  counters: (userId: number) =>
    apiFetch<RelationCountersResponse>(`${RELATION_PREFIX}/counter?userId=${userId}`)
};